lang/c++/test/SchemaTests.cc

```diff
@@ -658,6 +659,158 @@ static void testMalformedLogicalTypes(const char *schema) {
     BOOST_CHECK(datum.logicalType().type() == LogicalType::NONE);
 }
 
+static void testParseCustomAttributes() {
+    const std::string schema = R"({
+        "type": "record",
+        "name": "my_record",
+        "fields": [
+            { "name": "long_field",
+              "type": ["null", "long"],
+              "field-id": 1 },
+            { "name": "array_field",
+              "type": { "type": "array", "items": "int", "element-id": 3 },
+              "field-id": 2,
+              "extra": "1", "extra2": "2" },
+            { "name": "map_field",
+              "type": { "type": "map", "values": "int", "key-id": 5, "value-id": 6 },
+              "field-id": 4,
+              "extra": "foo" },
+            { "name": "timestamp_field",
+              "type": "long", "logicalType": "timestamp-micros", "adjust-to-utc": true,
+              "field-id": 10,
+              "extra": "bar" }
+        ]
+    })";
+
+    ValidSchema compiledSchema = compileJsonSchemaFromString(schema);
+    const NodePtr &root = compiledSchema.root();
+    BOOST_CHECK_EQUAL(root->customAttributes(), 4);
+
+    // long_field
+    {
+        auto customAttributes = root->customAttributesAt(0);
+        BOOST_CHECK_EQUAL(customAttributes.getAttribute("field-id").value(), "1");
+    }
+
+    // array_field
+    {
+        auto customAttributes = root->customAttributesAt(1);
+        BOOST_CHECK_EQUAL(customAttributes.getAttribute("extra").value(), "1");
```

Comment:
I find this API weird.
So, if the custom attributes are at the field's root level then you read them with `root->customAttributesAt(1).getAttribute("extra")`. But if there is there is something custom in the `type` then you need https://github.com/apache/avro/pull/3332/files#diff-559b3bdb2519c3e90855b7755c68a914b6507152dd6d1efe4fa77783ad4502c4R702-R705

I don't know what should be the correct behavior though. I think the Rust SDK just ignores the custom attributes in `type: {...}`

Reply:
This API aligns with its friends like `names` and `leaves` because they are for sub-fields: https://github.com/apache/avro/blob/main/lang/c%2B%2B/include/avro/Node.hh#L143-L160

Reply:
I agree that it is weird and this is different to the Java library where the custom attributes are directly accessed via the Node. However, the current implementation of `Node::addCustomAttributesForField()` already assigns one CustomAttribute to each leaf node. So we can only access the CustomAttribute of leaf node via Record/Map/Array's `customAttributesAt(i)` without breaking anything.